File: backend/app/routes/modules.py
from fastapi import APIRouter
from pydantic import BaseModel

from app.models.modules import AttachmentSocket, RigModule
from app.services import module_service

router = APIRouter(prefix="/api/modules", tags=["modules"])


class AssignModulesRequest(BaseModel):
    rig_job_id: str
    module_ids: list[str]


@router.get("", response_model=list[RigModule])
async def list_modules():
    return module_service.list_modules()


@router.get("/sockets", response_model=list[AttachmentSocket])
async def list_sockets():
    return module_service.list_sockets()


@router.post("/assign")
async def assign_modules(body: AssignModulesRequest):
    return module_service.assign_modules(body.rig_job_id, body.module_ids)


@router.get("/rig/{rig_job_id}")
async def get_rig_modules(rig_job_id: str):
    return module_service.get_rig_assignment(rig_job_id)

File: frontend/src/lib/modules.ts
export interface RigModule {
  id: string;
  name: string;
  module_type: string;
  description: string;
  compatible_rig_types: string[];
  required_bones: string[];
  created_at: string;
}

export interface AttachmentSocket {
  id: string;
  socket_name: string;
  parent_bone: string;
  position: [number, number, number];
  rotation: [number, number, number];
  scale: [number, number, number];
  attachment_type: string;
}

export interface TransformState {
  id: string;
  name: string;
  description: string;
  active_modules: string[];
  transform_behavior: string;
}

export interface ModuleAssignment {
  rig_job_id: string;
  assigned_modules: RigModule[];
  incompatible_modules: (RigModule & { incompatibility_reason: string })[];
  compatible_sockets: AttachmentSocket[];
  transform_states: TransformState[];
}

export async function getModules(): Promise<RigModule[]> {
  const res = await fetch("/api/modules");
  if (!res.ok) throw new Error("Failed to fetch modules");
  return res.json();
}

export async function getSockets(): Promise<AttachmentSocket[]> {
  const res = await fetch("/api/modules/sockets");
  if (!res.ok) throw new Error("Failed to fetch sockets");
  return res.json();
}

export async function assignModules(
  rig_job_id: string,
  module_ids: string[]
): Promise<ModuleAssignment> {
  const res = await fetch("/api/modules/assign", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ rig_job_id, module_ids }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(
      (err as { detail?: string }).detail ?? "Failed to assign modules"
    );
  }
  return res.json();
}

export async function getRigModules(
  rig_job_id: string
): Promise<ModuleAssignment> {
  const res = await fetch(`/api/modules/rig/${rig_job_id}`);
  if (!res.ok) throw new Error("Failed to fetch rig modules");
  return res.json();
}
